src/leaderboard/leaderboard.service.ts

```diff
@@ -0,0 +1,68 @@
+import { Injectable } from '@nestjs/common';
+import { ServiceHelper } from 'src/common/helpers/service.helper';
+import { LeaderboardRepository } from './repository/leaderboard.repository';
+import { LeaderboardDto } from './dto/leaderboard.dto';
+import { LeaderboardEntity as Leaderboard } from './entities/leaderboard.entity';
+import { PlayerLeaderboardRepository } from 'src/player-leaderboard/repository/player-leaderboard.repository';
+import { SortedResult } from './dto/sorted-result.dto';
+import { SortingOrders } from './entities/sorting.enum';
+import { PlayerLeaderboardEntity as PlayerLeaderboard } from 'src/player-leaderboard/entity/player-leaderboard.entity';
+
+@Injectable()
+export class LeaderboardService {
+  constructor(
+    private readonly serviceHelper: ServiceHelper,
+    private readonly leaderboardRepository: LeaderboardRepository,
+    private readonly playerLeaderboardRepository: PlayerLeaderboardRepository,
+  ) {}
+
+  async createLeaderboard(id: string | undefined, data: LeaderboardDto): Promise<Leaderboard> {
+    const fields: { [l: string]: any } = { ...data };
+    const newLeaderboard: Leaderboard = await this.serviceHelper.getUpsertData(id, fields, this.leaderboardRepository);
+    return this.leaderboardRepository.save(newLeaderboard);
+  }
+
+  async sortLeaderboard(
+    leaderboardId: string,
+    leaderboardName: string,
+    metrics: string[],
+    sortingOrder: SortingOrders[],
+    index: number,
+    list: PlayerLeaderboard[] | undefined,
+  ): Promise<SortedResult> {
+    if (list === undefined) {
+      const list = await this.playerLeaderboardRepository.find({
+        where: {
+          leaderboardId: leaderboardId,
+        },
+      });
+    }
+
+    if (sortingOrder[index].includes(SortingOrders.ASC)) {
```

Comment:
I'm not quite sure how to interpret the last part:
```
reverse = (sortingOrder === 'DESC') ? -1 : 1;

   if (a < b) result = reverse * -1;
   if (a > b) result = reverse * 1;
   if (result !== 0) `break;
```

But I made a first-try version with for loop:
```
async sortLeaderboard(
    leaderboardId: string,
  ): Promise<SortedResult> {

    let result: SortedResult;
    const leaderboard: Leaderboard = await this.leaderboardRepository.findOne(leaderboardId);
    const list = await this.playerLeaderboardRepository.find({
      where: {
        leaderboardId: leaderboardId,
      },
    });

    for (let i = 0; i < leaderboard.metrics.length; i++) {
      const metric = leaderboard.metrics[i];
      const sortingOrder = leaderboard.sortingOrders[i];
      const tempScores = [];

      if (sortingOrder === SortingOrders.ASC) {
        list.sort((x, y) => (x.score[metric] > y.score[metric] ? -1 : 1));
      }
      if (sortingOrder === SortingOrders.DESC) {
        list.sort((x, y) => (x.score[metric] > y.score[metric] ? 1 : -1));
      }

      for (const player of list) {
        tempScores.push(player.score[metric]);
      }

      if (!this.hasDuplicates(tempScores)) {
        result = {
          name: leaderboard.name,
          entries: list,
        };
        break;
      }

      return result;
    }

```

Reply:
The sample code I gave should come inside a `list.sort(...)`, and ideally you would not need anything else to sort except completing where it says compare.
```
  reverse = (sortingOrder === 'DESC') ? -1 : 1;  // do we need to reverse? then -1, otherwise 1

  if (a < b) result = reverse * -1; // if a is lower than b the compare function should give -1, but not if we want reverse (then multiply it by reverse value
  if (a > b) result = reverse * 1; // if a is greater than b the compare function should give 1, but not if we want reverse (then multiply it by reverse value
  if (result !== 0) break; // our values are not equal? great, we don't need to compare remaining metrics, exit loop
```

The following code inside your loop would re-sort the list for each metric.

```
if (sortingOrder === SortingOrders.ASC) {
  list.sort((x, y) => (x.score[metric] > y.score[metric] ? -1 : 1));
}
if (sortingOrder === SortingOrders.DESC) {
  list.sort((x, y) => (x.score[metric] > y.score[metric] ? 1 : -1));
}
```